Admin Wizard Choose License

leog opened the pull request:
## What does this PR do?

Implements 2 more steps in the Admin wizard for self-hosters to choose a license.

Closes #5998 #6014

**Environment**: Staging(main branch) / Production

Loom Video: UPDATED: https://loom.com/share/0d58e2fa5683478ebb52954b92eb6ee9

## Type of change

- [X] New feature (non-breaking change which adds functionality)

## How should this be tested?

Go to `/auth/setup` in a self-hosted Cal.com and see the wizard come up or If you don't have any user created in the database it will come up automatically when visiting the app. You can see two more steps or just one depending on the license chosen after setting up the first user (admin).

leog committed:
Implementation

leog commented on apps/web/pages/auth/setup/index.tsx:
```diff
@@ -9,26 +10,83 @@ import prisma from "@calcom/prisma";
+  const [isLoadingEnterpriseLicense, setIsLoadingEnterpriseLicense] = useState(false);
+  const [isEnabledEnterpriseLicense, setIsEnabledEnterpriseLicense] = useState(
+    !!props.deployment?.licenseKey
+  );
```
These are states to keep track of various things among steps.

zomars replied:
That's a lot of useStates

sean-brydon replied:
Tracking some of these in rhf or even into a single state object would probably be the best here 

From what i can see - a lot of these are controlled by what are effectively radio buttons and would work directly in rhf

leog replied:
Refactored to make the wizard handle it by the step content.

leog committed:
i18n

leog committed:
More i18n

leog commented on apps/web/pages/auth/setup/index.tsx:
```diff
@@ -9,26 +10,81 @@ import prisma from "@calcom/prisma";
+      ),
+      isLoading: isLoadingChooseLicense,
     },
+    ...(!isFreeLicense
```
If it's enterprise what's selected, add an additional step

leog commented on packages/config/tailwind-preset.js:
```diff
@@ -6,7 +6,7 @@ module.exports = {
     "./pages/**/*.{js,ts,jsx,tsx}",
     "./components/**/*.{js,ts,jsx,tsx}",
-    "../../packages/app-store/**/{components,pages}/**/*.{js,ts,jsx,tsx}",
+    "../../packages/app-store/**/{_components,_pages}/**/*.{js,ts,jsx,tsx}",
```
Components and pages folders were renamed and Tailwind process was not scanning them.

zomars replied:
This is an incomplete change. This is referring to each individual app's pages and components. You're replacing them with shared components. The correct solution should take both into account.

leog commented on packages/ui/components/form/step/Stepper.tsx:
```diff
@@ -16,12 +17,13 @@ function Stepper<T extends DefaultStep>(props: {
         <nav className="flex items-center justify-center" aria-label="Progress">
           <p className="text-sm font-medium">{stepLabel(props.step, steps.length)}</p>
-          <ol role="list" className="ml-8 flex items-center space-x-5">
+          <ol role="list" className="ml-8 flex items-center space-x-5" ref={stepperRef}>
```
Animating adding more steps dynamically.

leog commented on packages/ui/components/form/wizard/WizardForm.tsx:
```diff
@@ -37,7 +39,7 @@ function WizardForm<T extends DefaultStep>(props: {
       <div
         className={classNames(
-          "mb-8 divide-y divide-gray-200 overflow-hidden rounded-lg bg-white shadow print:divide-transparent print:shadow-transparent",
+          "mb-8 w-[700px] divide-y divide-gray-200 overflow-hidden rounded-lg bg-white shadow print:divide-transparent print:shadow-transparent",
```
Wizard was extending way more in width that it should, so limiting it from here.

zomars replied:
Should this be max width instead of a fixed one?

sean-brydon replied:
Yes - this will likely break on mobile without max width

sean-brydon commented on apps/web/pages/auth/setup/steps/ChooseLicense.tsx:
```diff
@@ -0,0 +1,70 @@
+            <li>{t("no_enterprise_features")}</li>
+          </ul>
+        </div>
+        <div
+          className={classNames(
+            "cursor-pointer space-y-2 rounded-md border p-4 hover:border-black",
+            !isFreeLicense && "border-black"
+          )}
```
Can we make this a button please

Clickable divs have no a11y support and require you to manually setup things like roles aria label and tab index aswell as a bunch of other things

sean-brydon commented on apps/web/pages/auth/setup/steps/ChooseLicense.tsx:
```diff
@@ -0,0 +1,70 @@
+        });
+      }}>
+      <div className="grid grid-cols-2 gap-4">
+        <div
+          className={classNames(
+            "cursor-pointer space-y-2 rounded-md border p-4 hover:border-black",
+            isFreeLicense && "border-black"
+          )}
+          onClick={() => setIsFreeLicense(true)}>
```
Can we make this a button please

Clickable divs have no a11y support and require you to manually setup things like roles aria label and tab index aswell as a bunch of other things

sean-brydon replied:
Also these appear to very simular to radio buttons under the hood - Use Radix if you need to do some fancy styling

sean-brydon commented on apps/web/pages/auth/setup/steps/ChooseLicense.tsx:
```diff
@@ -0,0 +1,70 @@
+    },
+  });
+  return (
+    <form
```
Use our Form component and use rhf to prevent some of the state tracking above

sean-brydon commented on apps/web/pages/auth/setup/steps/EnterpriseLicense.tsx:
```diff
@@ -0,0 +1,72 @@
+          className="mt-1"
+          placeholder="b6052119-1fee-42a6-90e3-55cjgjaaebb7"
+          labelSrOnly={true}
+          onChange={(e) => {
+            props.setIsEnabled(e.target.value !== "");
+            formMethods.setValue("licenseKey", e.target.value);
+          }}
+          id="licenseKey"
+          name="licenseKey"
```
```suggestion
          {...formMethods.register("licenseKey"),{
             onChange: (e) =>  props.setIsEnabled(e.target.value !== "")
          }}
```

sean-brydon committed:
Fix booking filters not working for admin (#6576)

zomars commented on packages/prisma/schema.prisma:
```diff
@@ -643,6 +653,11 @@ enum WorkflowMethods {
+  FREE
+  EE
+}
+
```
Can we split these migrations in a prep PR?

emrysal replied:
Do self hosters really care about these models?

zomars replied:
That's another good point @emrysal we don't want to mixup what should live in our console DB with here. We can also check for `CALCOM_LICENSE_KEY` for this.

leog replied:
The idea was to move out from environment variable, to let admins setup their instances as much as we can from the admin wizard, like we did with activating apps from it.

leog replied:
@zomars I'm afraid that doing a prep PR with the migrations separately could introduce churn on why was the new model was introduced. Any particular reason for doing it? Just to understand the rationale, learn and do it differently next time.

zomars replied:
Also we can assume that an empty key is a "FREE" license.

zomars commented on packages/app-store/_components/AppCategoryNavigation.tsx:
```diff
@@ -31,7 +31,7 @@ const AppCategoryNavigation = ({
           sticky
           linkProps={{ shallow: true }}
-          itemClassname={classNames(fromAdmin && "w-60")}
+          itemClassname={classNames(fromAdmin && "!w-48")}
```
what does the `!` do here?

leog replied:
Makes it `!important` in order to overwrite any other similar CSS.

zomars commented on packages/features/apps/AdminAppsList.tsx:
```diff
@@ -183,26 +185,36 @@ const querySchema = z.object({
     <form
-      id="wizard-step-2"
-      name="wizard-step-2"
+      id={`wizard-step-${currentStep}`}
+      name={`wizard-step-${currentStep}`}
```
Why do we need currentStep here other than naming these?

leog replied:
Apps list in the wizard can be step 3 or 4 depending on the license you choose, so it needs to be dynamically set.

zomars commented on packages/features/apps/AdminAppsList.tsx:
```diff
@@ -183,26 +185,36 @@ const querySchema = z.object({
       <AppCategoryNavigation
         baseURL={baseURL}
-        fromAdmin
+        fromAdmin={fromAdmin}
```
Is this being used elsewhere when fromAdmin is not true?

emrysal replied:
Yeah, `fromAdmin` is false on the /apps page. It's true from the Settings pages.

zomars commented on packages/prisma/migrations/20230119020155_deployment_model/migration.sql:
```diff
@@ -0,0 +1,14 @@
+-- CreateEnum
+CREATE TYPE "DeploymentLicenseType" AS ENUM ('FREE', 'EE');
+
+-- CreateTable
+CREATE TABLE "Deployment" (
+    "id" SERIAL NOT NULL,
+    "logo" TEXT,
+    "theme" JSONB,
+    "licenseKey" TEXT,
+    "licenseType" "DeploymentLicenseType",
+    "licenseConsentAt" TIMESTAMP(3),
+
+    CONSTRAINT "Deployment_pkey" PRIMARY KEY ("id")
+);
```
https://github.com/calcom/cal.com/pull/6574/files#r1081847900

zomars commented on packages/trpc/server/routers/viewer/auth.tsx:
```diff
@@ -86,6 +87,24 @@ export const authRouter = router({
+        ...(input.licenseKey ? { licenseKey: input.licenseKey } : {}),
+      };
+
+      await prisma.deployment.upsert({ where: { id: 1 }, create: data, update: data });
```
Since this is a special kind of table we should abstract these methods to limit the where clauses.  Since there will be only one and will be confusing for newcomers. could be something like `deploymentSettingsUpdate/Update` etc.

zomars reviewed (changes requested):
Blocking to address @sean-brydon and my feedback as well 🙏🏽

emrysal commented:
@leog could you add a loom 🙏 ?

leog commented:
> @leog could you add a loom 🙏 ?

@emrysal Totally forgot, will do.

> Do self hosters really care about these models?
> That's another good point @emrysal we don't want to mixup what should live in our console DB with here. We can also check for `CALCOM_LICENSE_KEY` for this.

@emrysal @zomars That's a good point, I just limited to implement what was needed from the ticket, which specifies creating those steps in the admin wizard and put things together following the `RFC: Self-hosted instance metadata DB table`. I was suggesting @PeerRich that we could probably send an API request to us to let us know when someone consents a license on their self-hosted instance to track down usages.

PeerRich commented:
how do I test this the best?

PeerRich commented on apps/web/public/static/locales/en/common.json:
```diff
@@ -1512,5 +1514,21 @@
+  "no_recordings_found": "No recordings found",
+  "agplv3_license": "AGPLv3 License",
+  "ee_enterprise_license": "“/ee” Enterprise License",
+  "enterprise_booking_fee": "$0.05/booking – min. $189/month",
```
I am fine merging this but we need to create an issue to later get this from our API endpoint because pricing will change eventually. We need an endpoint that always returns the most recent pricing string

PeerRich replied:
@zomars

PeerRich replied:
also, our pricing is about to change in the next 2 days anyway

leog commented:
> how do I test this the best?

Locally, removing all users if you want to go through the admin wizard from scratch or signing in as admin and going to `/auth/setup`

PeerRich committed:
extracted i18n, needs api to get most recent price, added hint: update later

PeerRich commented on apps/web/pages/auth/setup/steps/ChooseLicense.tsx:
```diff
@@ -0,0 +1,76 @@
+          </ul>
+        </div>
+      </div>
+      <div>
+        <small>{t("you_can_upgrade_at_any_time_later")}</small>
+      </div>
```
fyi: @leog added this section

PeerRich commented on apps/web/pages/auth/setup/steps/ChooseLicense.tsx:
```diff
@@ -0,0 +1,76 @@
+          )}
+          onClick={() => setIsFreeLicense(false)}>
+          <h2 className="font-cal text-xl text-black">{t("ee_enterprise_license")}</h2>
+          <p className="font-medium text-green-800">{t("enterprise_booking_fee", { enterprise_fee })}</p>
```
fyi: @leog made this a variable so we can pull it from an API endpoint to update later and not hardcode the price

PeerRich commented on apps/web/pages/auth/setup/steps/ChooseLicense.tsx:
```diff
@@ -0,0 +1,76 @@
+    },
+  });
+
+  const enterprise_fee = "$99"; // TODO: get this from a new API endpoint
```
@leog @zomars this needs to come from our backend so we can update them dynamically

PeerRich reviewed (approved):
submitted a commit, left some comments. rest looks good, very excited for this!

PeerRich commented:
oh i made a mistake, but should be an easy fix:
![CleanShot 2023-01-20 at 11 22 52@2x](https://user-images.githubusercontent.com/8019099/213672588-109d5722-7c57-45fd-acec-547a6ac579f7.jpg)

the {{enterprise_booking_fee}} is not being picked up correctly @leog

G3root committed:
fix: react-select overflow issue in some modals. (#6587)

* feat: add a disable overflow prop

* feat: use the disable overflow prop

sean-brydon committed:
Tailwind Merge (#6596)

* Tailwind Merge

* Fix merge classNames

gitstart-calcom committed:
[CAL-808] /availability/single - UI issue on buttons beside time inputs (#6561)

* [CAL-808] /availability/single - UI issue on buttons beside time inputs

* Update apps/web/public/static/locales/en/common.json

* Update packages/features/schedules/components/Schedule.tsx

* create new translation for tooltip

Co-authored-by: gitstart-calcom <gitstart@users.noreply.github.com>
Co-authored-by: Peer Richelsen <peeroke@gmail.com>
Co-authored-by: CarinaWolli <wollencarina@gmail.com>

zomars committed:
Bye bye submodules (#6585)

* WIP

* Uses ssh instead

* Update .gitignore

* Update .gitignore

* Update Makefile

* Update git-setup.sh

* Update git-setup.sh

* Replaced Makefile with bash script

* Update package.json

Udit-takkar committed:
fix: show button on empty string (#6601)

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

Udit-takkar committed:
fix: add delete in dropdown (#6599)

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

PeerRich committed:
Update README.md

PeerRich committed:
Update README.md

emrysal committed:
Changed a neutral- classes to gray (#6603)

* Changed a neutral- classes to gray

* Changed all border-1 to border

PeerRich committed:
Update package.json

zomars committed:
Test fixes

zomars committed:
Yarn lock fixes

leog committed:
Fixing i18n var

leog commented:
> the {{enterprise_booking_fee}} is not being picked up correctly @leog

@PeerRich Fixed. I will work on the suggestions from the other folks and will be ready Monday.

emrysal committed:
Fix string equality check in git-setup.sh

gitstart-calcom committed:
[CAL-811] Avatar icon not redirecting user back to the main page (#6586)

* Remove cursor-pointer, remove old Avatar* files

* Fixed styling for checkedSelect + some cleanup

Co-authored-by: gitstart-calcom <gitstart@users.noreply.github.com>
Co-authored-by: Alex van Andel <me@alexvanandel.com>

harshsinghatz committed:
Harsh/add member invite (#6598)

Co-authored-by: Guest <guest@pop-os.localdomain>
Co-authored-by: root <harsh.singh@gocomet.com>

emrysal committed:
Regenerated lockfile without upgrade (#6610)

luiisca committed:
fix: remove annoying outline when <Button /> clicked (#6537)

* fix: remove annoying outline when <Button /> clicked

* Delete yarn.lock

PeerRich committed:
remove 1 on 1 icon (#6609)

* removed 1-on-1 badge

* changed user to users for group events

G3root committed:
fix: case-sensitivity in apps path (#6552)

* fix: lowercase slug

* fix: make fallback blocking

sean-brydon committed:
Fix FAB (#6611)

Udit-takkar committed:
feat: add LocationSelect component (#6571)

* feat: add LocationSelect component

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

* fix: type error

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

* chore: type error

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

sean-brydon committed:
Update booking filters design  (#6543)

* Update booking filters

* Add filter on YOUR bookings

* Fix pending members showing up in list

emrysal committed:
Reduce the avatar size to 'sm' for now

emrysal committed:
Bugfix/dropdown menu trigger as child remove class names (#6614)

* Fix UsernameTextfield to take right height

* Remove className side-effect

* Incorrect resolution version fixed

* Converted mobile DropdownMenuTrigger styles into Button

emrysal committed:
v2.5.3

Udit-takkar committed:
fix: use items-center (#6618)

mocherfaoui committed:
fix tooltip and modal stacking issues (#6491)

* fix tooltip and modal stacking issues

* use z-index in larger screens and less

Co-authored-by: Alex van Andel <me@alexvanandel.com>

leog committed:
Temporary fix (#6626)

hariombalhara committed:
Fix Ga4 tracking (#6630)

sean-brydon committed:
generic <UpgradeScreen> component (#6594)

* first attempt of <UpgradeScreen>

* changes to icons

* reverted changes back to initial state, needs fix: teams not showing

* WIP

* Fix weird reactnode error

* Fix loading text

* added upgradeTip to routing forms

* icon colors

* create and use hook to check if user has team plan

* use useTeamPlan for upgradeTeamsBadge

* replace huge svg with compressed jpeg

* responsive fixes

* Update packages/ui/components/badge/UpgradeTeamsBadge.tsx

Co-authored-by: sean-brydon <55134778+sean-brydon@users.noreply.github.com>

* Give team plan features to E2E tests

* Allow option to make a user part of team int ests

* Remove flash of paywall for team user

* Add team user for typeform tests as well

Co-authored-by: Peer Richelsen <peer@cal.com>
Co-authored-by: CarinaWolli <wollencarina@gmail.com>
Co-authored-by: Carina Wollendorfer <30310907+CarinaWolli@users.noreply.github.com>
Co-authored-by: Alex van Andel <me@alexvanandel.com>
Co-authored-by: Hariom Balhara <hariombalhara@gmail.com>

leog committed:
Removing env var to rely on db

leog committed:
Merge branch 'main' into feat/admin-wizard-license

leog committed:
Merge branch 'main' into feat/admin-wizard-license

leog committed:
Restoring i18n keys, set loading moved

leog committed:
Merge branch 'main' into feat/admin-wizard-license

leog committed:
Fixing tailwind-preset glob

leog committed:
Wizard width fix for md+ screens

leog committed:
Converting licenses options to radix radio

leog committed:
Applying feedback + other tweaks

leog commented:
Latest Loom: https://loom.com/share/0d58e2fa5683478ebb52954b92eb6ee9

leog committed:
Merge branch 'main' into feat/admin-wizard-license

leog committed:
Reverting this, not this PR related

sean-brydon commented:
I'm not sure if its just me but i kinda believe this should be one form ultimately and we handle the final submit on the last step of the wizard and not on every step. 

UX seems weird to me if you start completing this (without ever hitting a final step in the form) - close the tab and you need to resume from where you left off. Im not sure if my perception of a stepper form is off  but this just went against my intuition when testing it out. 

We do this elsewhere - eventTypes is a good example ... Each page "settings tab" isnt a unique form and the content is wrapped in a single form itself. 

@emrysal @zomars maybe im just being nit picky here but this just didnt click with me 

I know [RHF](https://react-hook-form.com/advanced-usage/) has a good example of a form stepper where each step is contained within one form

leog commented on apps/web/pages/api/auth/[...nextauth].tsx:
```diff
@@ -331,10 +331,14 @@ export default NextAuth({
     async session({ session, token }) {
-      const hasValidLicense = await checkLicense(process.env.CALCOM_LICENSE_KEY || "");
+      const deployment = await prisma.deployment.findFirst({ where: { id: 1 } });
+      const hasValidLicense = await checkLicense(deployment?.licenseKey || "");
```
Now checking the new db model instead of environment variable

leog commented on apps/web/pages/api/link.ts:
```diff
@@ -39,7 +39,10 @@ async function handler(req: NextApiRequest, res: NextApiResponse<Response>) {
+      license: {
+        valid: true,
+        key: "",
+      },
```
License validity has been moved to its own object together with the key in order to have it at hand for other usages.

leog replied:
I needed the key in the telemetry context, perhaps we can remove it from telemetry so I can prescind from including it in the session, which I don't really like

leog commented on apps/web/pages/auth/setup/steps/StepDone.tsx:
```diff
@@ -1,20 +1,26 @@
+const StepDone = (props: {
+  currentStep: number;
+  nextStepPath: string;
+  setIsLoading: Dispatch<SetStateAction<boolean>>;
```
StepDone could be shown at any completed step so it was parametrized to accommodate any step.

leog commented on apps/web/public/static/locales/en/common.json:
```diff
@@ -836,7 +836,7 @@
   "create_events_on": "Create events on",
   "enterprise_license": "This is an enterprise feature",
-  "enterprise_license_description": "To enable this feature, get a deployment key at {{consoleUrl}} console and add it to your .env as CALCOM_LICENSE_KEY. If your team already has a license, please contact {{supportMail}} for help.",
+  "enterprise_license_description": "To enable this feature, have an administrator go to {{setupUrl}} to enter a license key. If a license key is already in place, please contact {{supportMail}} for help.",
```
Changed the wording given the fact we moved out from environment variable to the new db model through the admin wizard.

leog replied:
Not restoring this so we can rely on the admin wizard steps for people to have a better experience regarding license key

PeerRich replied:
this was a mistake. we need to keep `get a deployment key at {{consoleUrl}} console and`

leog commented on apps/web/public/static/locales/en/common.json:
```diff
@@ -1103,8 +1103,8 @@
-  "upgrade_to_access_recordings_description":"Recordings are only available as part of our teams plan. Upgrade to start recording your calls",
-  "recordings_are_part_of_the_teams_plan":"Recordings are part of the teams plan",
+  "upgrade_to_access_recordings_description": "Recordings are only available as part of our teams plan. Upgrade to start recording your calls",
+  "recordings_are_part_of_the_teams_plan": "Recordings are part of the teams plan",
```
i18n ally plugin corrections

leog commented on apps/web/public/static/locales/en/common.json:
```diff
@@ -1404,7 +1404,7 @@
   "team_url_taken": "This URL is already taken",
   "team_publish": "Publish team",
-  "number_sms_notifications": "Phone number (SMS\u00a0notifications)",
+  "number_sms_notifications": "Phone number (SMS notifications)",
```
i18n ally plugin corrections

leog commented on apps/web/public/static/locales/en/common.json:
```diff
@@ -1486,11 +1488,11 @@
   "no_availability_in_month": "No availability in {{month}}",
   "view_next_month": "View next month",
-  "send_code" : "Send code",
+  "send_code": "Send code",
```
i18n ally plugin corrections

leog commented on packages/config/tailwind-preset.js:
```diff
@@ -6,7 +6,7 @@ module.exports = {
     "./pages/**/*.{js,ts,jsx,tsx}",
     "./components/**/*.{js,ts,jsx,tsx}",
-    "../../packages/app-store/**/{components,pages}/**/*.{js,ts,jsx,tsx}",
+    "../../packages/app-store/**/*{components,pages}/**/*.{js,ts,jsx,tsx}",
```
Correcting the glob to take into account "_components" and "_pages" within app-store and "components" and "pages" within app folders.

leog commented on packages/features/apps/AdminAppsList.tsx:
```diff
@@ -183,26 +185,36 @@ const querySchema = z.object({
-        containerClassname="w-full xl:mx-5 xl:w-2/3 xl:pr-5"
+        containerClassname={classNames(
+          "w-full xl:mx-5 xl:w-2/3 xl:pr-3.5",
+          fromAdmin ? " max-h-[400px] overflow-y-auto" : "max-w-[500px]"
```
This is to have the app list within the admin wizard, contained inside the content section of the wizard and not expand its dimension, should be self-contained there.

leog commented on packages/features/ee/api-keys/components/ApiKeyListItem.tsx:
```diff
@@ -43,16 +43,8 @@ const ApiKeyListItem = ({
-            </Badge>
-          )}
+          {!neverExpires && isExpired && <Badge variant="red">{t("expired")}</Badge>}
+          {!isExpired && <Badge variant="green">{t("active")}</Badge>}
```
Testing the API Keys section of the settings to show the "needs license" warning, couldn't help to notice the badges were wrongly padded, this is the fix.

leog commented on packages/features/ee/common/components/LicenseRequired.tsx:
```diff
@@ -32,7 +32,7 @@ const LicenseRequired = ({ children, as = "", ...rest }: LicenseRequiredProps) =
   return (
     <Component {...rest}>
-      {session.data?.hasValidLicense ? (
+      {session.data?.license.valid ? (
```
License validity has been moved to its own object together with the key in order to have it at hand for other usages.

leog commented on packages/features/ee/common/components/v2/LicenseRequired.tsx:
```diff
@@ -18,7 +18,7 @@ const LicenseRequired = ({ children, as = "", ...rest }: LicenseRequiredProps) =
   const { t } = useLocale();
   const Component = as || Fragment;
-  const hasValidLicense = session.data ? session.data.hasValidLicense : null;
+  const hasValidLicense = session.data ? session.data.license.valid : null;
```
License validity within the session object has been moved to its own object together with the key in order to have it at hand for other usages.

leog commented on packages/features/ee/common/components/LicenseRequired.tsx:
```diff
@@ -46,6 +46,7 @@ const LicenseRequired = ({ children, as = "", ...rest }: LicenseRequiredProps) =
                     consoleUrl: `<a href="https://go.cal.com/console" target="_blank" class="underline">
                 ${APP_NAME}
               </a>`,
+                    setupUrl: `<a href="${WEBAPP_URL}/auth/setup">/auth/setup</a>`,
```
Text now refers to the admin wizard to be able to enter a license key instead of the environment variable.

leog commented on packages/features/ee/common/components/v2/LicenseRequired.tsx:
```diff
@@ -36,9 +36,9 @@ const LicenseRequired = ({ children, as = "", ...rest }: LicenseRequiredProps) =
                     consoleUrl: `<a href="${CONSOLE_URL}" target="_blank" rel="noopener noreferrer" class="underline">
                 ${APP_NAME}
               </a>`,
+                    setupUrl: `<a href="${WEBAPP_URL}/auth/setup" class="underline">/auth/setup</a>`,
```
Text now refers to the admin wizard to be able to enter a license key instead of the environment variable.

leog commented on packages/features/ee/common/components/v2/LicenseRequired.tsx:
```diff
@@ -36,9 +36,9 @@ const LicenseRequired = ({ children, as = "", ...rest }: LicenseRequiredProps) =
                     supportMail: `<a href="mailto:${SUPPORT_MAIL_ADDRESS}" class="underline">
-                ${SUPPORT_MAIL_ADDRESS}
-              </a>`,
+                ${SUPPORT_MAIL_ADDRESS}</a>`,
```
An extra underlined space after the email was shown, this fixes it.

leog commented on packages/features/ee/common/server/checkLicense.ts:
```diff
@@ -5,18 +5,9 @@ import { CONSOLE_URL } from "@calcom/lib/constants";
-    return v != null && v.length >= 2 && v.charAt(0) == '"' && v.charAt(v.length - 1) == '"'
-      ? v.substring(1, v.length - 1)
-      : v;
-  });
```
We went back on using `.uuid` directly as quotes are not an option when entering it in the admin wizard form.

leog commented on packages/features/ee/workflows/pages/index.tsx:
```diff
@@ -43,7 +43,7 @@ function WorkflowsPage() {
       subtitle={t("workflows_to_automate_notifications")}
       CTA={
-        session.data?.hasValidLicense && data?.workflows && data?.workflows.length > 0 ? (
+        session.data?.license.valid && data?.workflows && data?.workflows.length > 0 ? (
```
License validity within the session object has been moved to its own object together with the key in order to have it at hand for other usages.

leog commented on packages/features/ee/workflows/pages/workflow.tsx:
```diff
@@ -251,7 +251,7 @@ function WorkflowPage() {
         }
         heading={
-          session.data?.hasValidLicense &&
+          session.data?.license.valid &&
```
License validity within the session object has been moved to its own object together with the key in order to have it at hand for other usages.

leog committed:
Unneeded code removal

leog commented on packages/prisma/zod-utils.ts:
```diff
@@ -293,6 +293,24 @@ export const RoutingFormSettings = z
+    fontName: z.string().default("Cal Sans"),
+    fontSrc: z.string().default("https://cal.com/cal.ttf"),
+  })
+  .optional();
```
This is to support future theme customization within the admin wizard.

leog commented on packages/trpc/server/routers/viewer.tsx:
```diff
@@ -1148,6 +1148,7 @@ export const viewerRouter = mergeRouters(
     loggedInViewerRouter,
     public: publicViewerRouter,
     auth: authRouter,
+    deploymentSetup: deploymentSetupRouter,
```
As you suggested @zomars.

leog commented on packages/ui/components/button/Button.tsx:
```diff
@@ -210,7 +210,7 @@ export const Button = forwardRef<HTMLAnchorElement | HTMLButtonElement, ButtonPr
               className={classNames(
                 variant === "icon" && "h-4 w-4",
-                variant === "button" && "h-4 w-4 stroke-[1.5px] ltr:mr-2 rtl:ml-2"
+                variant === "button" && "h-4 w-4 stroke-[1.5px] ltr:ml-2 rtl:mr-2"
```
I think this was the other way around as a bug, could see it in the "Purchase a License" button as anchor that includes am "external link" icon at the right of the button text.

leog commented on packages/ui/components/button/Button.tsx:
```diff
@@ -245,7 +245,7 @@ export const Button = forwardRef<HTMLAnchorElement | HTMLButtonElement, ButtonPr
                 "inline-flex",
                 variant === "icon" && "h-4 w-4",
-                variant === "button" && "h-4 w-4 stroke-[1.5px] ltr:mr-2 rtl:ml-2"
+                variant === "button" && "h-4 w-4 stroke-[1.5px] ltr:ml-2 rtl:mr-2"
```
I think this was the other way around as a bug, could see it in the "Purchase a License" button as anchor that includes am "external link" icon at the right of the button text.

leog commented on packages/ui/components/credits/Credits.tsx:
```diff
@@ -23,7 +25,7 @@ export default function Credits() {
       {hasMounted && (
         <Link href="https://go.cal.com/releases" target="_blank" className="hover:underline">
-          {CalComVersion}
+          {CalComVersion(session.data?.license.valid)}
```
Relying on the license validity instead of its existence.

leog commented on packages/ui/components/form/inputs/Input.tsx:
```diff
@@ -117,7 +117,7 @@ export const InputField = forwardRef<HTMLInputElement, InputFieldProps>(function
       )}
       {addOnLeading || addOnSuffix ? (
-        <div className="relative mb-1 flex items-center rounded-md focus-within:outline-none focus-within:ring-2 focus-within:ring-neutral-800 focus-within:ring-offset-1">
+        <div className="group relative mb-1 flex items-center rounded-md focus-within:outline-none focus-within:ring-2 focus-within:ring-neutral-800 focus-within:ring-offset-1">
```
This allows to better control the hover states of the optional end or start icons of the field.

leog commented on packages/ui/components/form/wizard/WizardForm.tsx:
```diff
@@ -30,48 +33,52 @@ function WizardForm<T extends DefaultStep>(props: {
+
+  useEffect(() => {
+    setCurrentStepIsLoading(false);
+  }, [currentStep]);
```
This is to be able to delegate the `isLoading` state to the step itself if it includes the callable element for that purpose. With this I was able to remove all `useStates` you guys were noticing @zomars @sean-brydon.

zomars replied:
WDYM? I think I don't follow.

leog replied:
Instead of having an external state for each step which the step itself needs to manipulate and the wizard needs to know the result, I just made the step manipulate its own loading state and allowing the wizard access to it.

leog committed:
Reverting unneeded style change

leog commented on packages/ui/components/form/wizard/WizardForm.tsx:
```diff
@@ -30,48 +33,52 @@ function WizardForm<T extends DefaultStep>(props: {
     <div className="mx-auto mt-4 print:w-full">
       {/* eslint-disable-next-line @next/next/no-img-element */}
-      <img className="mx-auto mb-8 h-8" src="https://cal.com/logo.svg" alt="Cal.com Logo" />
+      <img className="mx-auto my-8 h-8" src="https://cal.com/logo.svg" alt="Cal.com Logo" />
```
Some space was needed also at the top of the Cal.com logo

sean-brydon reviewed (commented):
Also just a nit - i dont think this is the right stepper component? 

https://www.figma.com/file/9MOufQNLtdkpnDucmNX10R/%E2%9D%96-Cal-DS?node-id=1375%3A68066&t=MWFt6z1Mr9X67SHj-4

leog commented on packages/ui/components/form/wizard/WizardForm.tsx:
```diff
@@ -30,48 +33,52 @@ function WizardForm<T extends DefaultStep>(props: {
       <div
         className={classNames(
-          "mb-8 divide-y divide-gray-200 overflow-hidden rounded-lg bg-white shadow print:divide-transparent print:shadow-transparent",
+          "mb-8 divide-y divide-gray-200 overflow-hidden rounded-lg bg-white shadow print:divide-transparent print:shadow-transparent md:w-[700px]",
```
Just applying wizard form width correction on medium+ screens

leog commented:
> Also just a nit - i dont think this is the right stepper component?

We are using the same one that was already there for the wizard in Console if I remember correctly. You should definitely create a follow up issue to correct this if you like @sean-brydon, and let's confirm with @Jaibles just in case.

Jaibles commented:
We should be using the same stepper component everywhere. The one that's in [the DS](https://www.figma.com/file/9MOufQNLtdkpnDucmNX10R/%E2%9D%96-Cal-DS?node-id=1375%3A68066&t=skOTDLXzSf0UHM5W-1)

zomars commented:
How do FREE users finish this step?

![image](https://user-images.githubusercontent.com/3504472/214641114-29cd15d4-63f1-483e-93ee-181cf8003cc2.png)

zomars commented on apps/web/pages/auth/setup/steps/AdminUser.tsx:
```diff
@@ -2,6 +2,7 @@ import { zodResolver } from "@hookform/resolvers/zod";
 import classNames from "classnames";
```
These files should not live here:

![image](https://user-images.githubusercontent.com/3504472/214646092-758d3106-481b-4582-9a85-75bbf3bb3131.png)

These should not be valid URLs

http://localhost:3000/auth/setup/steps/AdminUser

zomars commented on apps/web/pages/auth/setup/index.tsx:
```diff
@@ -1,39 +1,86 @@
+      title: t("choose_a_license"),
+      description: t("choose_license_description"),
+      contentEl: (setIsLoading: Dispatch<SetStateAction<boolean>>) =>
+        props.deployment?.licenseType ? (
+          <StepDone currentStep={2} nextStepPath="/auth/setup?step=3" setIsLoading={setIsLoading} />
+        ) : (
```
What' s the point of having a "prev step" if we cannot do anything about it?

![image](https://user-images.githubusercontent.com/3504472/214646485-e8b9ccdf-9cfa-45bb-85c7-277fa1a72641.png)

IMO we should allow to go back and change the license. Makes sense to not do that with the first admin user creation tho.

leog replied:
Done!

zomars commented on packages/trpc/server/routers/viewer/deploymentSetup.tsx:
```diff
@@ -0,0 +1,27 @@
+  update: authedProcedure
+    .input(
+      z.object({
+        licenseType: z.nativeEnum(DeploymentLicenseType).optional(),
```
What is `licenseType` being used for?

leog replied:
to store the license type EE or FREE, if we choose to use `licenseKey` only, we can't distinguish between an admin that already went through the wizard or not

zomars replied:
How so? The wizard should display for empty license key

leog replied:
Ow, that's right, we don't need to remember which license was selected because we want to allow the admin to go back in the wizard to change it, making sense of the "Previous Step" button. Yeah, we can remove the license type I guess.

zomars commented on packages/types/environment.d.ts:
```diff
@@ -1,7 +1,5 @@
 declare namespace NodeJS {
   interface ProcessEnv {
-    /** Needed to enable enterprise-only features */
-    readonly CALCOM_LICENSE_KEY: string | undefined;
```
We should create a migration for upsetting the value to the DB first.

leog replied:
Did it here: https://github.com/calcom/cal.com/pull/6574/files#diff-da0845fc4b52d249c4f05f5150127d2d9bc00151a5a1272a69afa2b8a161bad9R114

zomars commented on packages/ui/components/credits/Credits.tsx:
```diff
@@ -1,15 +1,17 @@
-export const CalComVersion = `v.${pkg.version}-${!IS_SELF_HOSTED ? "h" : "sh"}-${
-  process.env.CALCOM_LICENSE_KEY === "" ? "ee" : "ce"
-}`;
+export const CalComVersion = (licenseKey: boolean | undefined) => {
+  return `v.${pkg.version}-${!IS_SELF_HOSTED ? "h" : "sh"}-${licenseKey ? "ee" : "ce"}`;
+};
```
This should come from a public endpoint

leog replied:
Is this a blocker? Otherwise I can work on it in a follow up PR

zomars commented on packages/ui/components/form/wizard/WizardForm.tsx:
```diff
@@ -1,16 +1,19 @@
-  enabled?: boolean;
-  isLoading: boolean;
+  content?: JSX.Element;
+  contentEl?: (setIsLoading: Dispatch<SetStateAction<boolean>>) => JSX.Element;
```
What does `contentEl` means in this context?

zomars reviewed (changes requested):
Overall nice progress @leog although I think this PR is not in a merge-able state yet. Left some feedback

zomars commented on packages/prisma/schema.prisma:
```diff
@@ -615,6 +615,16 @@ model WorkflowsOnEventTypes {
 }
 
+model Deployment {
+  id               Int                    @id @default(1)
```
This change is missing an extra migration. Will open up another PR to migrate first.

zomars replied:
https://github.com/calcom/cal.com/pull/6706

zomars commented on packages/lib/telemetry.ts:
```diff
@@ -35,24 +36,28 @@ export function collectPageParameters(
-      return fetch(url, { method: "POST", mode: "cors" });
-    } catch (e) {
-      console.error(`Error reporting booking for key: '${key}'`, e);
+  const session = await getSession();
```
Why use session instead of prisma here? AFAIK `getSession` makes a request as well so there's no fetching benefit here

leog replied:
I tried, but this code seems to be executed client side. I could rely on tRPC perhaps, or we can just remove the license key from telemetry all-together to avoid relying on session.

zomars commented on packages/lib/telemetry.ts:
```diff
@@ -35,24 +36,28 @@ export function collectPageParameters(
 export const nextCollectBasicSettings: CollectOpts = {
   drivers: [
-    process.env.CALCOM_LICENSE_KEY && process.env.NEXT_PUBLIC_IS_E2E !== "1" ? reportUsage : undefined,
+    reportUsage,
```
Do we want to report E2E usage?

leog replied:
No sure 🤔

leog commented on .env.example:
```diff
@@ -1,24 +1,10 @@
-# @see https://console.cal.com
-CALCOM_LICENSE_KEY=
-# ***********************************************************************************************************
-
```
Added a row creation instruction when visiting the admin setup to support the env var

leog committed:
Applying feedback

leog committed:
Removing licenseType

leog commented:
Applied all the feedback. Thanks everyone. @PeerRich @zomars @sean-brydon 

- Switched the stepper for the correct one
- Changed the wizard styles to match the design system
- Allowed to go back in steps without blocking choosing another option for license
- Removed the license type as we don't need to keep track of which license was selected to show already done steps
- Added "License" in Admin section of User settings and Cmd+k "Choose a license" option that gos to the admin wizard
- Dynamically removed the first step regarding creating the first (admin) user if user count was already different than 0
- Made sure console wizard also looked good with the wizard revamped styles to follow DS -> https://github.com/calcom/console/pull/64
- Also note there are changes in API to get the license key from the database instead of the env var -> https://github.com/calcom/api/pull/229

Here is what it looks like:

https://user-images.githubusercontent.com/467258/214700677-ac701a0c-b358-426c-9a9a-85e644dd804d.mov

PeerRich commented:
this is getting really close to completion! lots of great changes recently

sean-brydon commented:
> Applied all the feedback. Thanks everyone. @PeerRich @zomars @sean-brydon @Jaibles
> 
> * Switched the stepper for the correct one
> * Changed the wizard styles to match the design system
> * Allowed to go back in steps without blocking choosing another option for license
> * Removed the license type as we don't need to keep track of which license was selected to show already done steps
> * Added "License" in Admin section of User settings and Cmd+k "Choose a license" option that gos to the admin wizard
> * Dynamically removed the first step regarding creating the first (admin) user if user count was already different than 0
> * Made sure console wizard also looked good with the wizard revamped styles to follow DS -> [Revamped the wizard to accomodate new style console#64](https://github.com/calcom/console/pull/64)
> * Also note there are changes in API to get the license key from the database instead of the env var -> [Relying on deployment license key api#229](https://github.com/calcom/api/pull/229)
> 
> > Warning
> > There is still a pending discussion with @zomars related to having the license key in the session to be able to use it in telemetry.
> 
> Here is what admin wizard looks like now:
> 
>  Screen.Recording.2023-01-25.at.18.57.16.mov

Yoo this is looking good! Am i missing where the designs are for this or did @Jaibles pull a uno reverse card after the fact :D As these look a bit different https://www.figma.com/file/9MOufQNLtdkpnDucmNX10R/%E2%9D%96-Cal-DS?node-id=23849%3A147371&t=zWPdT0NqVfk8GfXA-4

leog commented:
> Yoo this is looking good! Am i missing where the designs are for this or did @Jaibles pull a uno reverse card after the fact :D As these look a bit different https://www.figma.com/file/9MOufQNLtdkpnDucmNX10R/%E2%9D%96-Cal-DS?node-id=23849%3A147371&t=zWPdT0NqVfk8GfXA-4

https://www.figma.com/file/UajFu4M1APhn2AywAEJkJr?node-id=1426:116568&comments-enabled=1&viewer=1 😖

sean-brydon commented:
> > Yoo this is looking good! Am i missing where the designs are for this or did @Jaibles pull a uno reverse card after the fact :D As these look a bit different https://www.figma.com/file/9MOufQNLtdkpnDucmNX10R/%E2%9D%96-Cal-DS?node-id=23849%3A147371&t=zWPdT0NqVfk8GfXA-4
> 
> https://www.figma.com/file/UajFu4M1APhn2AywAEJkJr?node-id=1426:116568&comments-enabled=1&viewer=1 😖

Lol nice - @Jaibles im gonna guess its the one in the DS file right??

@leog Happy to lend a hand refactoring the UI if it turns out to be the wrong one :D

Jaibles commented:
What @leog has done is perfect. There were old designs for the console there before. (Think even Peer or Bailey might have done them cause they did some console design awhile back but also could have been me along time ago.) I updated them today to match our current onboarding so it was easier to re-use our existing stepper component and so it's more visually inline. No major changes and either way would've been fine but looks great to me now.

zomars committed:
Upgrades typescript

zomars committed:
Update yarn lock

zomars committed:
Typings

hariombalhara committed:
Hotfix: ping,riverside,whereby and around not showing up in list (#6712)

hariombalhara committed:
Hotfix: ping,riverside,whereby and around not showing up in list (#6712) (#6713)

zomars committed:
Adds deployment settings to DB (#6706)

* WIP

* Adds DeploymentTheme

* Add missing migrations

* Adds client extensions for deployment

* Cleanup

zomars committed:
Merge branch 'main' into feat/admin-wizard-license

zomars committed:
Delete migration.sql

zomars committed:
Maximum call stack size exceeded fix?

leog committed:
Relying on both, env var and new model

leog committed:
Merge branch 'feat/admin-wizard-license' of github.com:calcom/cal.com into feat/admin-wizard-license

leog committed:
Restoring env example doc for backward compat

leog commented on apps/web/pages/auth/setup/index.tsx:
```diff
@@ -1,47 +1,98 @@
+            | F               | T             |        2 |
+            | F               | F             |        3 |
+          */
+          currentStep={(props.userCount === 0) === isFreeLicense ? 3 : !isFreeLicense ? 4 : 2}
```
Adding the rationale behind the condition just in case.

leog commented:
Thanks @Jaibles for the support 🤗

@zomars restored the env var for backward compatibility, the PR shrieked quite a bit now.

Other pending PRs related to this one:
 - Made sure console wizard also looked good with the wizard revamped styles to follow DS -> https://github.com/calcom/console/pull/64
 - Also note there are changes in API to get the license key from the database instead of the env var -> https://github.com/calcom/api/pull/229

zomars committed:
Merge branch 'main' into feat/admin-wizard-license

zomars committed:
Revert upgrade

zomars committed:
Update index.ts

zomars committed:
Delete index.ts

leog committed:
Merge branch 'main' into feat/admin-wizard-license

leog committed:
Merge branch 'main' into feat/admin-wizard-license

zomars committed:
Merge branch 'main' into feat/admin-wizard-license

zomars committed:
Self contained checkLicense

zomars commented on apps/web/pages/api/auth/[...nextauth].tsx:
```diff
@@ -360,7 +360,9 @@ export default NextAuth({
     },
     async session({ session, token }) {
-      const hasValidLicense = await checkLicense(process.env.CALCOM_LICENSE_KEY || "");
+      const deployment = await prisma.deployment.findFirst({ where: { id: 1 } });
+      const licenseKey = process.env.CALCOM_LICENSE_KEY || deployment?.licenseKey || "";
+      const hasValidLicense = await checkLicense(licenseKey);
```
Will update to self-contain. Also for public API.

zomars commented on apps/web/pages/auth/setup/index.tsx:
```diff
@@ -62,9 +119,21 @@ export const getServerSideProps = async (context: GetServerSidePropsContext) =>
   return {
     props: {
       userCount,
+      deployment,
```
We should never expose the licenseKey to the frontend. This will.

leog replied:
Fixed, thx!

zomars commented on apps/web/pages/auth/setup/index.tsx:
```diff
@@ -1,47 +1,100 @@
   const { t } = useLocale();
-  const [isLoadingStep1, setIsLoadingStep1] = useState(false);
-  const shouldDisable = props.userCount !== 0;
+  const [isFreeLicense, setIsFreeLicense] = useState(props.deployment?.licenseKey === "");
+  const [isEnabledEE, setIsEnabledEE] = useState(!!props.deployment?.licenseKey);
```
We should not read the licenseKey in the frontend. Also if the license KEY is set we cannot choose the FREE version without any explanation. We should allow to select FREE or remove it altogether. It creates unexpected behavior.

leog replied:
Fixed, thx!

zomars commented on apps/web/components/setup/ChooseLicense.tsx:
```diff
@@ -0,0 +1,78 @@
+
+import { useLocale } from "@calcom/lib/hooks/useLocale";
+
+const ChooseLicense = (props: {
+  currentStep: number;
+  isFreeLicense: boolean;
+  setIsLoading: Dispatch<SetStateAction<boolean>>;
+  setIsFreeLicense: Dispatch<SetStateAction<boolean>>;
+}) => {
```
This components is caring for way more that it needs to:

```suggestion
const ChooseLicense = (
  props: {
    value: string;
    onChange: (value: string) => void;
    onSubmit: (value: EventTarget) => void;
  } & JSX.IntrinsicElements["form"]
) => {
```

leog replied:
Actually, we need the `currentStep` to be able to let the form know which step is the submit button expecting to submit.

zomars replied:
Refactored to reduce prop drilling.

zomars replied:
> Actually, we need the `currentStep` to be able to let the form know which step is the submit button expecting to submit.

We only need ´currenStep´ to form the new href. This can happen on the parent hence the use of onSubmit instead.

leog committed:
Not exposing license key, fixed radio behavior

leog committed:
Covering undefined env var

leog commented on apps/web/pages/auth/setup/index.tsx:
```diff
@@ -62,9 +119,21 @@ export const getServerSideProps = async (context: GetServerSidePropsContext) =>
   }
 
+  // Check existant CALCOM_LICENSE_KEY env var and acccount for it
+  if (process.env.CALCOM_LICENSE_KEY !== "" && !deployment?.licenseKey) {
```
Covered, thx!

zomars committed:
Feedback

zomars committed:
Moar feedback

zomars commented on packages/ui/components/form/step/Steps.tsx:
```diff
@@ -5,25 +5,32 @@ interface ISteps {
 
 const Steps = (props: ISteps) => {
-  const { maxSteps, currentStep, navigateToStep } = props;
   const { t } = useLocale();
```
Following the Stepper component example

zomars committed:
Feedback

zomars committed:
Feedback

zomars committed:
Feedback

zomars committed:
Cleanup

zomars reviewed (approved):
![LGTM](https://media2.giphy.com/media/OrFmkOFx7PVK/200.gif)

leog commented:
We should have tested this better before merging @zomars, I had to create #6954 and #6977 afterwards.
